File: others/分类/二叉树/987. 二叉树的垂序遍历.cpp
#include <algorithm>
#include <queue>
#include <map>
#include <functional>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

using namespace std;

class Solution {
public:
    vector<vector<int>> verticalTraversal(TreeNode *root) {
        map<int, map<int, vector<int>>> columnMap;

        function<void(TreeNode *, int, int)> dfs = [&](TreeNode *node, int row, int col) {
            if (!node) return;
            columnMap[col][row].push_back(node->val);
            dfs(node->left, row + 1, col - 1);
            dfs(node->right, row + 1, col + 1);
            };

        dfs(root, 0, 0);

        vector<vector<int>> result;
        for (auto &[col, rowMap] : columnMap) {
            vector<int> colNodes;
            for (auto &[row, values] : rowMap) {
                sort(values.begin(), values.end());
                colNodes.insert(colNodes.end(), values.begin(), values.end());
            }
            result.push_back(colNodes);
        }

        return result;
    }
};

File: leetcode/900-1000/987. 二叉树的垂序遍历.py
# Definition for a binary tree node.
from collections import defaultdict
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def verticalTraversal(self, root: Optional[TreeNode]) -> List[List[int]]:
        ans = [[] for _ in range(2005)]

        def dfs(node: Optional[TreeNode], x, y):
            if not node:
                return
            ans[y + 1002].append((x, node.val))
            dfs(node.left, x + 1, y - 1)
            dfs(node.right, x + 1, y + 1)

        dfs(root, 0, 0)
        real_ans = []
        for i in range(2005):
            if ans[i]:  # 如果当前列有值
                col = sorted(ans[i])  # 按 x、val 排序
                real_ans.append([val for x, val in col])
        return real_ans


class Solution:
    def verticalTraversal(self, root: Optional[TreeNode]) -> List[List[int]]:
        groups = defaultdict(list)

        def dfs(node: Optional[TreeNode], row: int, col: int):
            if node is None:
                return
            groups[col].append((row, node.val))  # col 相同的分到同一组
            dfs(node.left, row + 1, col - 1)
            dfs(node.right, row + 1, col + 1)

        dfs(root, 0, 0)

        ans = []
        for _, g in sorted(groups.items()):
            g.sort()  # 按照 row 排序，row 相同按照 val 排序
            ans.append([val for _, val in g])
        return ans
